test/common/quic/envoy_quic_h3_fuzz_helper.h

```diff
@@ -0,0 +1,98 @@
+#include <set>
+
+#include "test/common/quic/envoy_quic_h3_fuzz.pb.h"
+
+#include "quiche/quic/core/crypto/null_decrypter.h"
+#include "quiche/quic/core/crypto/null_encrypter.h"
+#include "quiche/quic/core/quic_connection.h"
+#include "quiche/quic/core/quic_versions.h"
+
+namespace Envoy {
+namespace Quic {
+
+// This class serializes structured protobuf `HTTP/3` messages to bytes as they
+// would be sent over the wire.
+class H3Serializer {
+public:
+  H3Serializer(std::set<uint32_t>& streams) : open_h3_streams_(streams){};
+  size_t serialize(char* buffer, size_t buffer_len, bool unidirectional, uint32_t type, uint32_t id,
+                   const test::common::quic::H3Frame& h3frame);
+
+private:
+  std::set<uint32_t>& open_h3_streams_;
+};
+
+// This class serializes structured protobuf `QUIC + HTTP/3` messages to bytes as they
+// would be sent over the wire.
+class QuicPacketizer {
+public:
+  struct QuicPacket {
+    size_t size{0};
+    char payload[1460];
+  };
+
+  QuicPacketizer(const quic::ParsedQuicVersion& quic_version,
+                 quic::QuicConnectionHelperInterface* connection_helper);
+  size_t serializePackets(const test::common::quic::QuicH3FuzzCase& input, QuicPacket* packets,
+                          size_t max_packets);
+  void reset();
+
+private:
+  bool serializePacket(const test::common::quic::QuicFrame& frame, QuicPacket* packet);
+  bool serializeJunkPacket(const std::string& data, QuicPacket* packet);
+  void serialize(quic::QuicFrame, QuicPacket* packet);
+  void serializeStreamFrame(const test::common::quic::QuicStreamFrame& frame, QuicPacket* packet);
+  void serializeNewTokenFrame(const test::common::quic::QuicNewTokenFrame& frame,
+                              QuicPacket* packet);
+  void serializeMessageFrame(const test::common::quic::QuicMessageFrame& frame, QuicPacket* packet);
+  void serializeCryptoFrame(const test::common::quic::QuicCryptoFrame& frame, QuicPacket* packet);
+  void serializeAckFrame(const test::common::quic::QuicAckFrame& frame, QuicPacket* packet);
+  void serializeNewConnectionIdFrame(const test::common::quic::QuicNewConnectionIdFrame& frame,
+                                     QuicPacket* packet);
+
+  quic::ParsedQuicVersion quic_version_;
+  quic::QuicConnectionHelperInterface* connection_helper_;
+  quic::QuicPacketNumber packet_number_;
+
+  quic::QuicConnectionId destination_connection_id_;
+  quic::QuicFramer framer_;
+
+  H3Serializer h3serializer_;
+  // Unidirectional streams are started by sending a varint indicating the stream type.
+  // This set tracks which stream ids have already been established so that the stream type
+  // is not inserted into a previously opened stream. RFC9114, Sec. 6.2
```

Comment:
How about:

// Unidirectional streams are started by sending a variable-length integer indicating the stream type (RFC9114, Sec. 6.2). H3Serializer serializes the stream type into the stream payload upon seeing the stream for the first time.
// This set tracks opened unidirectional streams in this flight.  So that multiple stream frames on the same stream will not cause the stream type to be serialized again.